Refactors and preparatory works for having tabbed navigation in stream edit modal.

m-e-l-u-h-a-n opened the pull request:
This PR does following refactors as  preparatory works for having tabbed design stream edit page. It contains commits from @pragatiagrawal31's #15445 . We do following changes commit-wise:

* Use consistent structure using divs instead of `li`s and `ul` for organizing content in stream edit page.
* Showing stream deactivation button at the end.
* Cleanup subscription.css to remove dead CSS rules from it.
* Extract modal related css rules to a separate modal.css files.
**Commits left for merging:**
* Remove content-editable widget for updating stream name and description and add a modal for handling these updates.
* Wrap contents into different settings group (General | Personal | Subscribers).
* Add tabbed widget for navigation between different settings in stream settings page.

**Testing plan:** Mostly manually, and node tests.

**GIFs or screenshots:** 
![tabbed_design](https://user-images.githubusercontent.com/63504956/123514840-7deae180-d6b2-11eb-8045-69c21dc64869.gif)

m-e-l-u-h-a-n commented:
@timabbott before proceeding to make concrete changes for tabbed design for stream edit page it would be better to get reviews on this refactor as further work if done separately will conflict with(so I am currently stacking on top of these commits).

m-e-l-u-h-a-n commented on frontend_tests/puppeteer_tests/settings.ts:
```diff
@@ -355,7 +355,6 @@ async function test_default_language_setting(page: Page): Promise<void> {
     await page.goto("http://zulip.zulipdev.com:9981/#settings"); // get back to normal language.
     await page.waitForSelector(display_settings_section, {visible: true});
     await page.click(display_settings_section);
-    await page.waitForSelector("#language-settings-status", {visible: true});
```
We already check that save alert indicator appears on line:354. So no need to assert its visibility again as it is hidden by default(when it has no content).

timabbott replied:
Makes sense.  As a side note, in English parentheticals should always have a space before them, so:

> by default(when it has no content).

should be

> by default (when it has no content).

It's not particularly important, but I have been finding it distracting whenever I read it, so I think it is worth tweaking your writing style :)

timabbott commented on static/templates/stream_creation_form.hbs:
```diff
@@ -1,6 +1,6 @@
 <div class="hide" id="stream-creation" tabindex="-1" role="dialog"
   aria-label="{{t 'Stream creation' }}">
-    <form id="stream_creation_form" class="form-inline">
+    <form id="stream_creation_form">
```
What does `form-inline` do?  It appears to be a bootstrap thing.

m-e-l-u-h-a-n replied:
It's a bootstrap class. It's most common use is in setting display properties to various input elements in the form so that if there is sufficient space horizontally input elements are shown like:
`label: input element`.
But as we override almost all of the styles set by its class (checked using developer tools), there is no effect on design by removing it.
I'll move this to a separate PR along with other places in stream-edit page where I have checked it is can be safely removed.

timabbott commented on static/templates/stream_types.hbs:
```diff
@@ -1,43 +1,49 @@
               class="stream_message_retention_setting" class="prop-element"
               {{#if disable_message_retention_setting}}disabled{{/if}}>
-                <option value="realm_default">{{t 'Use organization level settings' }} {{realm_message_retention_setting}}</option>
+                <option value="realm_default">{{t "Use organization level settings "}}{{realm_message_retention_setting}} </option>
```
What's the story with these changes?  A screenshot might be helpful.

In any case, I think the i18n approach taken here seems wrong, since we seem to be relying on word order being like English; I think what we want is `"Use organization level settings ({realm_message_retention_setting})"` as the string we provide to translators, so they can adjust word order.

m-e-l-u-h-a-n replied:
Fixed translation issue in its own separate commit. Also I have modified the commit message to explain more clearly about the reason for html changes in this commit ([https://github.com/zulip/zulip/pull/18872/commits/4eaf45f1ac51aaa62db1e54683bc7838b948e098](https://github.com/zulip/zulip/pull/18872/commits/4eaf45f1ac51aaa62db1e54683bc7838b948e098)).

timabbott commented on static/templates/change_stream_info_modal.hbs:
```diff
@@ -0,0 +1,34 @@
+            <span aria-hidden="true">&times;</span>
+        </button>
+        <h3 id="change_stream_info_label">
+            {{t "Change stream info for #" }}{{ stream_name }}
```
Same comment on translation tags and word order here.  See https://zulip.readthedocs.io/en/latest/translating/internationalization.html for background.

m-e-l-u-h-a-n replied:
Made this change. Also added an extra commit for fixing similar issue in `subscription_stream_privacy_modal.hbs`.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -789,12 +699,61 @@ export function initialize() {
+        (e) => {
+            // This fixes a weird bug in which, subscription_settings hides
+            // unexpectedly by clicking the cancel button in a modal on top of it.
+            e.stopPropagation();
```
I feel like we shouldn't need something like this -- it feels like the bug is that the modal backdrop isn't working properly.

Perhaps @sahil839 or @ganpa3 can find the proper fix for this.

m-e-l-u-h-a-n replied:
There is one more instance of this same problem in this file.
[https://github.com/zulip/zulip/blob/054be8dd2e18a381d601200fc1d5198889ccd7a6/static/js/stream_edit.js#L790](https://github.com/zulip/zulip/blob/054be8dd2e18a381d601200fc1d5198889ccd7a6/static/js/stream_edit.js#L790)

ganpa3 replied:
The modal closes due to a click handler registered in [here](https://github.com/zulip/zulip/blob/777eeb6b8f551c0c0fdeffb2334a595b40fb73f0/static/js/subs.js#L999). Either changing that (not ideal) or appending the `change_stream_info_modal` to `body` will fix this naturally.

timabbott replied:
Can't we just tighten that selector to be `foo[data-dismiss]` for the appropriate value of `foo` that is the `data-dismiss` element for the "Stream settings" overlay?

m-e-l-u-h-a-n replied:
Yeah, Tim's suggestion about this seems to work perfectly fine for this.
I tried to find a bit more about why that line was added in `subs.js` and found that it was specifically added for the `cancel` button in stream creation form in [https://github.com/zulip/zulip/commit/1886f0a015c5c3a9141a8cbd439d0c737dde2af8](https://github.com/zulip/zulip/commit/1886f0a015c5c3a9141a8cbd439d0c737dde2af8). I think it assumed that we will have a data-dismiss for that button only inside `.subscription` but as we now have new modals (and corresponding `data-dismiss` elements for them) due to stream-edit section. So binding that selector correctly appeared like a natural fix for it.
I added a commit (https://github.com/zulip/zulip/pull/18872/commits/6084bf980f5891bc474d76476a261b1e0810e43a) just next to this so that both commits are coherent.

ganpa3 replied:
> Can't we just tighten that selector to be `foo[data-dismiss]` for the appropriate value of `foo` that is the `data-dismiss` element for the "Stream settings" overlay?

Yeah, that's what I suggested.

timabbott commented on static/templates/change_stream_info_modal.hbs:
```diff
@@ -0,0 +1,34 @@
+            {{t "Save changes"}}
+        </button>
+    </div>
+</div>
```
This is great as an intermediate state, but ultimately we probably want this to be part of the same top-level UI widget as the stream's configuration (Basically similar to an organization settings page, with multiple sections, and name+description being one of them).  I think we gain a lot by just removing the `content-editable stuff`.

m-e-l-u-h-a-n replied:
Organization level settings are mostly dropdown select widgets. For this may be we take inspiration from from how we allow users to edit their name/email (May be discuss this in detail after tabbed design migration 🤔 ).

timabbott replied:
Yeah, I think this is a fine thing to discuss once we've done that cutover.

timabbott commented on static/styles/subscriptions.css:
```diff
@@ -882,50 +869,6 @@
-        &:hover {
-            color: hsl(0, 0%, 27%);
-        }
-    }
```
How carefully have you audited that we found all of the old `editable` CSS?

m-e-l-u-h-a-n replied:
I used `git grep -n "editable\|contenteditable" static/styles/` and after this removal we only have `contenteditable` related css for settings pages where input pills are used like user grorup settings modal.

timabbott commented on static/styles/settings.css:
```diff
@@ -520,19 +520,20 @@ input[type="checkbox"] {
-    margin-left: 10px;
-    color: hsl(156, 30%, 50%);
-    padding: 3px 10px;
+    display: none;
```
Can you add a comment repeating some of the commit message notes here?  And maybe also show a screenshot of what the border looks like; we may want to backport this commit to 4.x if it's really ugly.

m-e-l-u-h-a-n replied:
On checking for screenshot, I cannot reproduce this. But I remember seeing empty alert box, after adding a basic tabbed navigation widget for trial. I have dropped this commit now but I will add it again if it appears again after adding tabbed navigation.

timabbott replied:
OK.

timabbott commented on static/styles/modal.css:
```diff
@@ -0,0 +1,37 @@
+
+.modal-bg {
+    background-color: hsl(0, 0%, 98%);
+}
```
This is a nice extraction!

timabbott commented:
@m-e-l-u-h-a-n thanks for rebasing this!  I'm really excited about this refactor, and completely agree we should try to get it in before doing the big migration of stream settings to a tabbed UI.  I posted a bunch of comments.  

The logic changes in this PR are somewhat hard for me to read/verify, and I think would have been better done with some of the more complex commits split into multiple smaller commits, but it's possible it's not worth the effort to do so.  That said, a lot of the smaller changes that I posted questions about seem potentially independent of the rest of the commit, and the way I'd typically write a PR is to put those in their own mini commits with a commit message answering the natural question of why we're making that change.

Anyway, thanks for working on this!  I've been wanting to kill those content-editable widgets for years :)

m-e-l-u-h-a-n commented:
@timabbott I have updated this pr and replied to changes done for each comment in their respective threads. I added separate commits for translation issues (as they felt unrelated to rest of the commit).

For `.form-inline` thing I wish to open separate pr removing that class from stream edit/creation page wherever possible.
It is ready for another review.

m-e-l-u-h-a-n commented on static/templates/stream_types.hbs:
```diff
@@ -43,7 +43,7 @@
             <select name="stream_message_retention_setting"
               class="stream_message_retention_setting" class="prop-element"
               {{#if disable_message_retention_setting}}disabled{{/if}}>
-                <option value="realm_default">{{t 'Use organization level settings' }} {{realm_message_retention_setting}}</option>
+                <option value="realm_default">{{#tr}}Use organization level settings {org_level_message_retention_setting}{{/tr}}</option>
```
Changed this template variable name to `org_level_message_retention_setting` because it gave errors in CI because of word realm present in previous name.

timabbott replied:
I think this is a case where a linter exclude rule would be OK, but this is a fine solution.

timabbott commented:
@m-e-l-u-h-a-n can you rebase this and add a summary comment on what (if anything) this still needs before merging?  This is a reasonably complex project, and I'd like to push hard on integrating it so that we can start planning follow-ups.  I'm really excited about this!

m-e-l-u-h-a-n commented:
I was waiting for https://github.com/zulip/zulip/pull/18872#discussion_r653136084 to be resolved and added a commit (https://github.com/zulip/zulip/pull/18872/commits/6084bf980f5891bc474d76476a261b1e0810e43a) for this. I think this does not require any further additions, and as the next step after this, I am planning to add the tabbed design because this contains almost all of the refactors that were needed to have a neat migration to tabbed design.

timabbott commented:
I merged all but the `content_editable` cutover commit as the series ending with c36aa1126c72fbf55e2e746c46d4b629db8e6b6e, after doing somewhat extensive QA on the user experience with just all the refactoring commits included (the rebase was clean).  

I didn't merge the cutover because the animation for saving a change to a stream name is super weird (in that it creates a row with some weird green flash behavior). One theory for how to fix this is that we perhaps want the "Saving" item to live in the row containing the stream name and "Preview" widget (and just have the stream name get abbreviated with `...` temporarily if space is needed via overflow CSS).

But it's also possible that this problem gets indirectly addressed by the tab-switcher migration commit, so if it's annoying to fix, we can potentially just wait to merge this last commit until we can do so together with that cutover.

m-e-l-u-h-a-n commented:
I've updated this PR with changes that implement tabbed navigation. It contains following three changes now:
* Remove content-editable widget for updating stream name and description and add a modal for handling these updates.
* Wrap contents into different settings group (General | Personal | Subscribers).
* Add tabbed widget for navigation between different settings in stream settings page.

About saving indicator problem I've asked for some ideas at https://chat.zulip.org/#narrow/stream/101-design/topic/stream.20settings.20page/near/1219211.

timabbott commented on static/templates/subscription_settings.hbs:
```diff
@@ -2,46 +2,46 @@
+                    </div>
+                    <a href="{{preview_url}}" class="button small rounded" id="preview-stream-button" role="button" title="{{t 'View stream'}} (V)" {{#unless should_display_preview_button }}style="display: none"{{/unless}}><i class="fa fa-eye"></i></a>
+                    {{#if is_realm_admin}}
+                    <button class="button small rounded btn-danger deactivate" type="button" name="delete_button" title="{{t 'Delete stream'}}"> <i class="fa fa-trash-o" aria-hidden="true"></i></button>
```
This s/Archive/Delete/` change here is wrong.

m-e-l-u-h-a-n replied:
Oops, fixed now.

timabbott commented on static/templates/subscription_settings.hbs:
```diff
@@ -1,8 +1,23 @@
+        </div>
+        <a href="{{preview_url}}" class="button small rounded" id="preview-stream-button" role="button" title="{{t 'View stream'}} (V)" {{#unless should_display_preview_button }}style="display: none"{{/unless}}><i class="fa fa-eye"></i></a>
+        {{#if is_realm_admin}}
+        <button class="button small rounded btn-danger deactivate" type="button" name="delete_button" title="{{t 'Delete stream'}}"> <i class="fa fa-trash-o" aria-hidden="true"></i></button>
```
This should be called "Archive stream"; we changed that recently and I think you accidentally reverted that change when resolving merge conflicts.

timabbott commented:
This is great. I think we can likely merge the first two commits once we:
* [ ] Fix that "Saving" widget issue
* [ ] Change the title for the modal from "Change stream info for #announce" to "Edit #announce".

Once that's done, I think a next step that could be done in a separate PR that we merge quickly is this:
* [ ] Merge the "Change" button modal with the "Edit stream" modal, so that they're just one edit-general-stream-settings widget.

There's a bundle of UI changes that I think we should stack as individual commits on top of this migration to make this mergeable without being a clear UI regression:

On the common tabs system:
* [ ] Increase the font size for the tab titles to match that for a section in that non-stream "Settings" UI.
* [ ] Change the "Stream settings" in the top-left corner to use the `Settings for #{stream_name}`, and use the same font size we use for that area in "Manage streams" (but probably not small caps).  We will likely want to iterate on this further after merging, but we should display clearly which stream is selected somewhere in the right side region.

On the "Subscribers" tab:
* [ ] Change the title to "Subscribers" from "Stream membership"
* [ ] Move the "Email address" block to the bottom of "General", since that's where it belongs.
* [ ] Change the users region to fill the full height of the page, rather than being fixed height.

On the "Personal" tab:
* [ ] Make the design identical to how a normal "settings / organization settings" page works, i.e. remove the box, dividers, etc.

On the "General" tab:
* [ ] Show text like "This stream does not have a description configured" if no description is set.
* [ ] Replace the "This is a public stream. Any member of the organization can join without an invitation.Only organization administrators can post." text, which now feels weirdly compressed given all the available space, with something that explains each value on a line by itself in a normal font size. A bulleted list might work as a first pass; we should try that and see if we need to do something harder.

Does that plan make sense to you?

m-e-l-u-h-a-n commented:
I've updated this pull request with following changes:
* Fixed save widget issue.
* change title in stream edit modal.
**Editing a stream having longest possible name:**
![save_widget](https://user-images.githubusercontent.com/63504956/124140475-2380c500-daa6-11eb-8ca1-6c7b34965797.gif)


I have moved the tabbed design commit in a different branch so that this PR can be merged. As this thread is getting cluttered I'll open a separate PR with tabbed deign commit and other suggestions (https://github.com/zulip/zulip/pull/18872#issuecomment-870984616) on top of that.

timabbott commented:
I took care of the small rebase issue and confirmed the fixes for the last batch of comments.  Merged, thanks @m-e-l-u-h-a-n!

I'm really excited to see this project progressing; we've been talking about it for forever :).
